### PR TITLE
Process OAuth errors to the UI page

### PR DESCRIPTION
**Pull request by mshaposhnik**
Signed-off-by: Max Shaposhnik <mshaposh@redhat.com>

### What does this PR do?
Process OAuth errors to the meaningful UI page.

### Screenshot/screencast of this PR

image

### How to test this PR?
- Use `mshaposh/spis:latest`
 - Setup OAuth flow, but decline authorization

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -39,9 +39,37 @@ type cliArgs struct {
...
+	data := viewData{
+		Title:   errorMsg,
+		Message: errorDescription,
+	}
+	tmpl, _ := template.ParseFiles("static/callback_error.html")
+
+	w.WriteHeader(http.StatusOK)
+	err := tmpl.Execute(w, data)
+	if err != nil {
+		zap.L().Error("failed to process template: %s", zap.Error(err))
```
Comment (skabashnyuk) [inline main.go:68]: should we return 500 here?

Comment (skabashnyuk) [reply main.go:68]: also. Should we update the state of the Token here?

**Review by skabashnyuk — COMMENTED** [review]
In general looks good

**Comment by skabashnyuk** [comment]
Got this error on my side.

```
ok  	github.com/redhat-appstudio/service-provider-integration-oauth	0.846s
?   	github.com/redhat-appstudio/service-provider-integration-oauth/authentication	[no test files]
Running Suite: SPI Oauth Controller Integration Test Suite
==========================================================
Random Seed: 1645431661
Will run 3 of 3 specs

STEP: bootstrapping test environment
2022-02-21T10:21:01.909+0200	DEBUG	controller-runtime.test-env	using existing cluster
2022-02-21T10:21:01.909+0200	DEBUG	controller-runtime.test-env	automatically acquiring client configuration
2022-02-21T10:21:02.352+0200	DEBUG	controller-runtime.test-env	installing CRDs
Failure [30.451 seconds]
[BeforeSuite] BeforeSuite
/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/suite_test.go:53

  Unexpected error:
      <*fmt.wrapError | 0xc00007c5a0>: {
          msg: "unable to install CRDs onto control plane: unable to create CRD instances: unable to create client: Get \"https://192.168.64.3:8443/api?timeout=32s\": dial tcp 192.168.64.3:8443: i/o timeout",
          err: <*fmt.wrapError | 0xc00007c580>{
              msg: "unable to create CRD instances: unable to create client: Get \"https://192.168.64.3:8443/api?timeout=32s\": dial tcp 192.168.64.3:8443: i/o timeout",
              err: <*fmt.wrapError | 0xc00007c560>{
                  msg: "unable to create client: Get \"https://192.168.64.3:8443/api?timeout=32s\": dial tcp 192.168.64.3:8443: i/o timeout",
                  err: <*url.Error | 0xc0004b3230>{
                      Op: "Get",
                      URL: "https://192.168.64.3:8443/api?timeout=32s",
                      Err: <*net.OpError | 0xc0004000a0>{
                          Op: "dial",
                          Net: "tcp",
                          Source: nil,
                          Addr: <*net.TCPAddr | 0xc0004b31d0>{
                              IP: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 192, 168, 64, 3],
                              Port: 8443,
                              Zone: "",
                          },
                          Err: <*net.timeoutError | 0x3393e00>{},
                      },
                  },
              },
          },
      }
      unable to install CRDs onto control plane: unable to create CRD instances: unable to create client: Get "https://192.168.64.3:8443/api?timeout=32s": dial tcp 192.168.64.3:8443: i/o timeout
  occurred

  /Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/suite_test.go:82
------------------------------
STEP: tearing down the test environment
Panic [0.000 seconds]
[AfterSuite] AfterSuite
/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/suite_test.go:149

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/local/Cellar/go/1.17.6/libexec/src/runtime/panic.go:221

  Full Stack Trace
  github.com/redhat-appstudio/service-provider-integration-oauth/controllers.glob..func3()
  	/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/suite_test.go:152 +0x4f
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0x1fa4b9a)
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/leafnodes/runner.go:113 +0xba
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run(0xc0002919f8)
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/leafnodes/runner.go:64 +0x125
  github.com/onsi/ginkgo/internal/leafnodes.(*simpleSuiteNode).Run(0xc0003575e0, 0x10080c5, 0xc000466e40, {0x100ebb4, 0xc000478ee0})
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/leafnodes/suite_nodes.go:25 +0x86
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runAfterSuite(0xc00033de40)
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/specrunner/spec_runner.go:138 +0xc7
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run(0xc00033de40)
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/specrunner/spec_runner.go:71 +0xdd
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run(0xc00073f1f0, {0xb856920, 0xc000484ea0}, {0x240cebf, 0x1}, {0xc000740bd0, 0x1, 0x1}, {0x2629278, 0xc000744740}, ...)
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/internal/suite/suite.go:79 +0x4d2
  github.com/onsi/ginkgo.runSpecsWithCustomReporters({0x25e2c00, 0xc000484ea0}, {0x240cebf, 0x2b}, {0xc000503720, 0x1, 0x10564df})
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/ginkgo_dsl.go:245 +0x185
  github.com/onsi/ginkgo.RunSpecs({0x25e2c00, 0xc000484ea0}, {0x240cebf, 0x2b})
  	/Users/skabashn/dev/src/go/pkg/mod/github.com/onsi/ginkgo@v1.16.5/ginkgo_dsl.go:220 +0x14a
  github.com/redhat-appstudio/service-provider-integration-oauth/controllers.TestSuite(0x1007979)
  	/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/suite_test.go:50 +0x90
  testing.tRunner(0xc000484ea0, 0x24a1550)
  	/usr/local/Cellar/go/1.17.6/libexec/src/testing/testing.go:1259 +0x102
  created by testing.(*T).Run
  	/usr/local/Cellar/go/1.17.6/libexec/src/testing/testing.go:1306 +0x35a
------------------------------

Ran 3 of 0 Specs in 30.452 seconds
FAIL! -- 0 Passed | 3 Failed | 0 Pending | 0 Skipped
--- FAIL: TestSuite (30.45s)
FAIL
FAIL	github.com/redhat-appstudio/service-provider-integration-oauth/controllers	31.591s
FAIL

```

**Comment by mshaposhnik** [comment]
@skabashnyuk AFAIK those tests are running under K8S, so you need an running minikube for them

**Comment by skabashnyuk** [comment]
```
❯ go test ./...
ok  	github.com/redhat-appstudio/service-provider-integration-oauth	(cached)
?   	github.com/redhat-appstudio/service-provider-integration-oauth/authentication	[no test files]
Running Suite: SPI Oauth Controller Integration Test Suite
==========================================================
Random Seed: 1645439946
Will run 3 of 3 specs

•Cleanup startedCleanup finished•Cleanup startedCleanup finished
------------------------------
• Failure [0.021 seconds]
Controller
/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/common_test.go:43
  when OAuth initiated
  /Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/common_test.go:142
    redirects to specified url [It]
    /Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/common_test.go:207

    Expected
        <int>: 500
    to equal
        <int>: 302

    /Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/common_test.go:244
------------------------------


Summarizing 1 Failure:

[Fail] Controller when OAuth initiated [It] redirects to specified url
/Users/skabashn/dev/src/redhat-appstudio/service-provider-integration-oauth/controllers/common_test.go:244

Ran 3 of 3 Specs in 1.114 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestSuite (1.11s)
FAIL
FAIL	github.com/redhat-appstudio/service-provider-integration-oauth/controllers
```

File: main.go (hunk 1 above)
Comment (mshaposhnik) [reply main.go:69]: imo it's better to return 500 + at least the textual representation of the error if we cannot parse template

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**